patch file
diff --git a/nixos/modules/system/boot/stage-1-init.sh b/nixos/modules/system/boot/stage-1-init.sh
index 59cf1a47f..9acbd7733 100644
--- a/nixos/modules/system/boot/stage-1-init.sh
+++ b/nixos/modules/system/boot/stage-1-init.sh
@@ -1,5 +1,7 @@
 #! @shell@
 
+set -x
+
 targetRoot=/mnt-root
 console=tty1
 verbose="@verbose@"
@@ -76,6 +78,25 @@ info
 info "[1;32m<<< @distroName@ Stage 1 >>>[0m"
 info
 
+info
+info "[1;32mlisting contents of /[0m"
+info
+
+ls -l --color /
+
+info
+info "[1;32mlisting contents of /dev[0m"
+info
+
+ls -l --color /dev
+
+info
+info "[1;32mdropping to bash shell[0m"
+info "[1;33mtype 'exit' to continue booting[0m"
+info
+
+bash || true
+
 # Make several required directories.
 mkdir -p /etc/udev
 touch /etc/fstab # to shut up mount
@@ -149,7 +170,7 @@ specialMount() {
   local fsType="$4"
 
   mkdir -m 0755 -p "$mountPoint"
-  mount -n -t "$fsType" -o "$options" "$device" "$mountPoint"
+  mount -v -n -t "$fsType" -o "$options" "$device" "$mountPoint"
 }
 source @earlyMountScript@
 
diff --git a/nixos/modules/system/boot/stage-2-init.sh b/nixos/modules/system/boot/stage-2-init.sh
index a89e3d817..3b46cb650 100755
--- a/nixos/modules/system/boot/stage-2-init.sh
+++ b/nixos/modules/system/boot/stage-2-init.sh
@@ -1,5 +1,7 @@
 #! @shell@
 
+set -x
+
 systemConfig=@systemConfig@
 
 export HOME=/root PATH="@path@"
@@ -79,6 +81,24 @@ if [ -n "@readOnlyNixStore@" ]; then
     fi
 fi
 
+echo
+echo "[1;32mlisting contents of /[0m"
+echo
+
+ls -l --color /
+
+echo
+echo "[1;32mlisting contents of /dev[0m"
+echo
+
+ls -l --color /dev
+
+echo
+echo "[1;32mdropping to bash shell[0m"
+echo "[1;33mtype 'exit' to continue booting[0m"
+echo
+
+bash || true
 
 if [ "${IN_NIXOS_SYSTEMD_STAGE1:-}" != true ]; then
     # Use /etc/resolv.conf supplied by systemd-nspawn, if applicable.
@@ -142,6 +162,7 @@ if [ "${IN_NIXOS_SYSTEMD_STAGE1:-}" != true ]; then
 
 
     # Start systemd in a clean environment.
-    echo "starting systemd..."
-    exec @systemdExecutable@ "$@"
+    echo "we would do 'starting systemd...' but instead we shall start bash"
+    #exec @systemdExecutable@ "$@"
+    exec bash
 fi
